src/transformers/models/timesfm/modeling_timesfm.py

```diff
@@ -0,0 +1,562 @@
+        )
+        f_emb = self.freq_emb(freq)  # B x 1 x D
+        model_input += f_emb
+
+        print(">>> PatchedDecoder model_input", model_input.shape)
+        model_output, all_attentions, all_hidden_states = self.stacked_transformer(model_input, patched_padding, output_attentions=output_attentions, output_hidden_states=output_hidden_states)
+        if output_hidden_states:
+            all_hidden_states = [model_input] + all_hidden_states
+
+        output_ts = self._postprocess_output(model_output, num_outputs, stats)
+        return output_ts, all_attentions, all_hidden_states
+
+    def decode(
+        self,
+        input_ts: torch.Tensor,
+        paddings: torch.Tensor,
+        freq: torch.LongTensor,
+        horizon_len: int,
+        output_patch_len: int | None = None,
+        max_len: int = 512,
+        return_forecast_on_context: bool = False,
+        output_attentions: bool = False,
+        output_hidden_states: bool = False,
+    ):
+        """Auto-regressive decoding without caching.
+
+        Args:
+          input_ts: input time-series and paddings. Time-series shape B x C.
+          paddings: padding shape B x (C + H) where H is the prediction length.
+          freq: frequency shape B x 1
+          horizon_len: prediction length.
+          output_patch_len: output length to be fetched from one step of
+            auto-regressive decoding.
+          max_len: maximum training context length.
+          return_forecast_on_context: whether to return the model forecast on the
+            context except the first input patch.
+
+        Returns:
+          Tuple of two forecasting results:
+          - Point (mean) output predictions as a tensor with shape B x H'.
+          - Full predictions (mean and quantiles) as a tensor with shape
+            B x H' x (1 + # quantiles).
+          In particular, if return_forecast_on_context is True, H' is H plus
+          the forecastable context length, i.e. context_len - (first) patch_len.
+        """
+        final_out = input_ts
+        context_len = final_out.shape[1]
+        full_outputs = []
+        if paddings.shape[1] != final_out.shape[1] + horizon_len:
+            raise ValueError(
+                "Length of paddings must match length of input + horizon_len:"
+                f" {paddings.shape[1]} != {final_out.shape[1]} + {horizon_len}"
+            )
+        if output_patch_len is None:
+            output_patch_len = self.config.horizon_len
+        num_decode_patches = (horizon_len + output_patch_len - 1) // output_patch_len
+        for step_index in range(num_decode_patches):
+            current_padding = paddings[:, 0 : final_out.shape[1]]
+            input_ts = final_out[:, -max_len:]
+            input_padding = current_padding[:, -max_len:]
+            fprop_outputs, all_attentions, all_hidden_states = self.forward(input_ts, input_padding, freq, output_attentions=output_attentions, output_hidden_states=output_hidden_states)
```

Comment:
Is caching being used to save on key value computations here? May be that is the purpose of `output_attentions` and `output_hidden_states` ?

Reply:
It is required by the Hugging Face framework.